backend/src/router.rs

```diff
@@ -0,0 +1,504 @@
+use crate::prelude::*;
+use axum::body::HttpBody;
+use axum::routing::on;
+use axum::{handler::Handler, routing::MethodFilter, Router};
+use hyper::{Body, Method};
+use std::convert::Infallible;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use thiserror::Error;
+use utoipa::openapi::PathItemType;
+use utoipa::OpenApi;
+
+#[derive(Debug, Error, PartialEq, Eq, PartialOrd, Ord)]
+pub enum RouteError {
+    #[error("No route found in OpenAPI scheme")]
+    NoRoute,
+    #[error("No method found for specified route in OpenAPI scheme")]
+    NoMethod,
+    #[error("No `operation_id` found")]
+    NoOperation,
+    #[error("OpenAPI's `operation_id` doesn't match handler's name")]
+    NoMatch,
+    #[error("Unexpected Hyper method - was it `Method::CONNECT`?")]
+    UnexpectedMethod,
+    #[error("Error occured during handler processing")]
+    InvalidHandler,
+}
+
+pub struct NoApi;
+
+impl<'a> ApiVersion<'a> for NoApi {}
+
+pub struct ApiV1;
+
+pub trait ApiVersion<'a> {
+    #[must_use]
+    fn to_path() -> &'a str {
+        ""
+    }
+}
+
+impl<'a> ApiVersion<'a> for ApiV1 {
+    fn to_path() -> &'a str {
+        "/api/v1"
+    }
+}
+
+pub struct ContextRouter<Version, Doc, S = (), B = Body> {
+    inner: Router<S, B>,
+    context: PhantomData<(Version, Doc)>,
+}
+
+impl<'a, Version, Doc, S, B> ContextRouter<Version, Doc, S, B>
+where
+    Version: ApiVersion<'a>,
+    Doc: OpenApi,
+{
+    pub fn no_context(self) -> Router<S, B> {
+        self.inner
+    }
+
+    #[must_use]
+    pub fn new_context<V, D>(self) -> ContextRouter<V, D, S, B> {
+        ContextRouter {
+            inner: self.inner,
+            context: PhantomData,
+        }
+    }
+
+    /// Calls `api_route` on inner `Router`
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error in the same cases as inner `api_route` call
+    pub fn api_route<H, T>(
+        mut self,
+        path: &str,
+        method: Method,
+        handler: H,
+    ) -> Result<Self, RouteError>
```

Comment:
It is better to aggregate errors inside in a vector, and return them from `no_context`. This is better for two reasons: 
1. there is no need to write `expect()` after each registration
2. at the end there will be a message with all the errors at once and there is no need to catch them one by one

Reply:
`error_stack`'s `Report` structure has [`extend_one` method](https://docs.rs/error-stack/latest/error_stack/struct.Report.html#method.extend_one). I think it's better to return the single report (with all of the errors) instead of vector with errors.

Reply:
Please, check how the final error is printed. All aggregated errors should be at one logical level. If they are, then you can use `extend_one`. It should not be the case that the first registered error became the base error for the second one and so on.

Reply:
they are on the same level